gto/git_utils.py

```diff
@@ -0,0 +1,71 @@
+import inspect
+import logging
+import re
+from functools import wraps
+from tempfile import TemporaryDirectory
+from typing import Callable, Dict
+
+from git import Repo
+
+# TODO: make a new type out of Union[str, Repo]
+
+
+def git_clone_if_repo_is_remote(f: Callable):
+    @wraps(f)
+    def wrapped_f(*args, **kwargs):
+        kwargs = _turn_args_into_kwargs(args, kwargs)
+
+        # noinspection PyTypeChecker
+        if isinstance(kwargs["repo"], str) and is_url_to_remote_repo(
+            repo=kwargs["repo"]
+        ):
+            try:
+                with TemporaryDirectory() as tmp_dir:
+                    logging.debug("create temporary directory %s", tmp_dir)
+                    # noinspection PyTypeChecker
+                    git_clone(repo=kwargs["repo"], dir=tmp_dir)
+                    kwargs["repo"] = tmp_dir
+                    result = f(**kwargs)
+            except (NotADirectoryError, PermissionError) as e:
+                raise e.__class__(
+                    "Are you using windows with python < 3.9? "
+                    "This may be the reason of this error: https://bugs.python.org/issue42796. "
+                    "Consider upgrading python."
+                ) from e
+            logging.debug("temporary directory %s has been deleted", tmp_dir)
+        else:
+            result = f(**kwargs)
+
+        return result
+
+    def _turn_args_into_kwargs(
+        args: tuple, kwargs: Dict[str, object]
+    ) -> Dict[str, object]:
+        kwargs_complement = {
+            k: args[i]
+            for i, k in enumerate(inspect.getfullargspec(f).args)
+            if k not in kwargs.keys() and i < len(args)
+        }
+        kwargs.update(kwargs_complement)
+        return kwargs
+
+    return wrapped_f
```

Comment:
This should work since repo is always the first arg to all CLI functions, and is more concise:

```suggestion
def git_clone_if_repo_is_remote(f: Callable):
    @wraps(f)
    def wrapped_f(repo, *args, **kwargs):
        if isinstance(repo, str) and is_url_to_remote_repo(repo=repo):
            try:
                with TemporaryDirectory() as tmp_dir:
                    logging.debug("create temporary directory %s", tmp_dir)
                    git_clone(repo=repo, dir=tmp_dir)
                    return f(tmp_dir, *args, **kwargs)
            except (NotADirectoryError, PermissionError) as e:
                raise e.__class__(
                    "Are you using windows with python < 3.9? "
                    "This may be the reason of this error: https://bugs.python.org/issue42796. "
                    "Consider upgrading python."
                ) from e
            logging.debug("temporary directory %s has been deleted", tmp_dir)
        else:
            return f(tmp_dir, args, **kwargs)

    return wrapped_f
```

Reply:
Mmm... I wrote the tests to make sure that the decorator works even if repo is not the first argument. I find it very much prone to problems if you create a decorator with such assumption.

So, currently your code suggestion breaks the tests even after fixing line 30 to `return f(repo, *args, **kwargs)`.

Are you sure you want to make the decorator less general? How would you suggest to make clear that the repo argument must be the first argument in the decorated function?

Reply:
Ok, I see. Let's keep it this then!